### ISSUE TITLE
Multinode Inference using deepspeed zero3

### ISSUE THREAD

**Issue by NotTheStallion** [huggingface/accelerate#2913]
### System Info

```Shell
- `Accelerate` version: 0.31.0
- Platform: Linux-6.5.0-41-generic-x86_64-with-glibc2.35
- `accelerate` bash location: /home/ubuntu/Myenv/bin/accelerate
- Python version: 3.10.12
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.3.1+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- System RAM: 31.08 GB
- GPU type: NVIDIA GeForce RTX 3060
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: DEEPSPEED
        - use_cpu: False
        - debug: True
        - num_processes: 7
        - machine_rank: 0
        - num_machines: 7
        - main_process_ip: main-node
        - main_process_port: 5000
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - deepspeed_config: {'deepspeed_config_file': '/home/ubuntu/_local_cuisine/inference_config.json', 'deepspeed_exclusion_filter': None, 'deepspeed_hostfile': '/home/ubuntu/_local_cuisine/hostfile', 'deepspeed_inclusion_filter': 'localhost:0@worker-01:0@worker-02:0@worker-03:0@worker-08:0@worker-09:0@worker-15:0', 'deepspeed_multinode_launcher': 'pdsh', 'zero3_init_flag': True}
        - downcast_bf16: yes
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I am trying to use the zero3 stage of deepspeed to distribute the model params and do inference using multiple machines (nodes) 

This is the code i am using (modified and simplified version of the example in the repo) :

      ``import argparse
      from transformers import AutoModel, pipeline, logging, AutoTokenizer, AutoModelForCausalLM, GPT2TokenizerFast, GPT2LMHeadModel, set_seed
      from safetensors import safe_open
      from safetensors.torch import save_file
      import evaluate
      import torch
      from datasets import load_dataset
      from torch.optim import AdamW
      from torch.utils.data import DataLoader
      from transformers import AutoModelForSequenceClassification, AutoTokenizer, get_linear_schedule_with_warmup, set_seed
      
      from accelerate import Accelerator, DistributedType
      
      MAX_GPU_BATCH_SIZE = 16
      EVAL_BATCH_SIZE = 32
      repo_id = "eminAydin/Turkish-GPT2-mini-ModelD"
      
      
      def get_dataloaders(accelerator: Accelerator, batch_size: int = 16):
          global tokenizer
          tokenizer = GPT2TokenizerFast.from_pretrained(repo_id)
          datasets = load_dataset("glue", "mrpc")
          
          # Apply the method we just defined to all the examples in all the splits of the dataset
          # starting with the main process first:
          with accelerator.main_process_first():
              tokenized_datasets = datasets.map(
                  lambda x:x,
                  batched=True,
                  remove_columns=["idx", "sentence1", "sentence2"],
              )
      
          # We also rename the 'label' column to 'labels' which is the expected name for labels by the models of the
          # transformers library
          tokenized_datasets = tokenized_datasets.rename_column("label", "labels")
          return tokenized_datasets
      
      
      def inference_function(config, args):
          global tokenizer
          # Initialize accelerator
          accelerator = Accelerator()
          seed = int(config["seed"])
      
          set_seed(seed)
      
          # Instantiate the model (we build the model here so that the seed also control new weights initialization)
          model = GPT2LMHeadModel.from_pretrained(repo_id, use_safetensors=True, return_dict=True)
      
          # We could avoid this line since the accelerator is set with `device_placement=True` (default value).
          # Note that if you are placing tensors on devices manually, this line absolutely needs to be before the optimizer
          # creation otherwise training will not work on TPU (`accelerate` will kindly throw an error to make us aware of that).
          model = model.to(accelerator.device)
      
          import torch
          from torch.utils.data import Dataset, DataLoader
      
          # Sample string
          data_string = "selam"
          data_list = data_string.split()
          class StringDataset(Dataset):
              def __init__(self, data):
                  self.data = data
      
              def __len__(self):
                  return len(self.data)
      
              def __getitem__(self, idx):
                  return self.data[idx]
      
          dataset = StringDataset(data_list)
          dataloader = DataLoader(dataset, batch_size=1, shuffle=True)
      
          # Prepare everything
          # There is no specific order to remember, we just need to unpack the objects in the same order we gave them to the
          # prepare method.
      
          model, dataloader = accelerator.prepare(
              model, dataloader
          )
      
          model.train()
          for step, batch in enumerate(dataloader):
              # We could avoid this line since we set the accelerator with `device_placement=True`.
              print(f"===============batch=============== {batch}")
              inputs = tokenizer(batch)
              print(f"===============inputs=============== {inputs}")
              inputs = torch.tensor(inputs["input_ids"][:1]).to(accelerator.device)
              outputs = model.generate(inputs)
              print(f"==============output================ {outputs}")
              response = tokenizer.decode(outputs[0])
      def main():
          parser = argparse.ArgumentParser(description="Simple example of training script.")
          parser.add_argument(
              "--mixed_precision",
              type=str,
              default=None,
              choices=["no", "fp16", "bf16", "fp8"],
              help="Whether to use mixed precision. Choose"
              "between fp16 and bf16 (bfloat16). Bf16 requires PyTorch >= 1.10."
              "and an Nvidia Ampere GPU.",
          )
          parser.add_argument("--cpu", action="store_true", help="If passed, will train on the CPU.")
          args = parser.parse_args()
          config = {"seed": 42}
          inference_function(config, args)
      if __name__ == "__main__":
          main()``

Command `accelerate launch ./_local_cuisine/inference_deep.py` This code creates 414MB in each GPU knowing that the entire model is 164MB.

This code does the inference of the prompt on all devices and loads the model in all GPUs.
Thus giving me this output `localhost: ===============batch=============== ['selam']
worker-01: ===============batch=============== ['selam']
worker-02: ===============batch=============== ['selam']
worker-03: ===============batch=============== ['selam']
worker-08: ===============batch=============== ['selam']
worker-09: ===============batch=============== ['selam']
worker-01: ===============inputs=============== {'input_ids': [[19706]], 'attention_mask': [[1]]}
worker-02: ===============inputs=============== {'input_ids': [[19706]], 'attention_mask': [[1]]}
worker-15: ===============batch=============== ['selam']
worker-03: ===============inputs=============== {'input_ids': [[19706]], 'attention_mask': [[1]]}
localhost: ===============inputs=============== {'input_ids': [[19706]], 'attention_mask': [[1]]}
worker-08: ===============inputs=============== {'input_ids': [[19706]], 'attention_mask': [[1]]}
worker-09: ===============inputs=============== {'input_ids': [[19706]], 'attention_mask': [[1]]}
worker-15: ===============inputs=============== {'input_ids': [[19706]], 'attention_mask': [[1]]}
worker-01: /home/ubuntu/Myenv/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
worker-01:   warnings.warn(
worker-02: /home/ubuntu/Myenv/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
worker-02:   warnings.warn(
worker-09: /home/ubuntu/Myenv/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
worker-09:   warnings.warn(
worker-15: /home/ubuntu/Myenv/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
worker-15:   warnings.warn(
worker-03: /home/ubuntu/Myenv/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
worker-03:   warnings.warn(
worker-08: /home/ubuntu/Myenv/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
worker-08:   warnings.warn(
localhost: /home/ubuntu/Myenv/lib/python3.10/site-packages/transformers/generation/utils.py:1168: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
localhost:   warnings.warn(
localhost: ==============output================ tensor([[19706,    12,   745,   646,   411,  1171,   307,   922,    14,   199,
localhost:             12,   400,   646,  1171,   307,   922,    14,   199,    12,   400]],
localhost:        device='cuda:0')
worker-08: ==============output================ tensor([[19706,    12,   745,   646,   411,  1171,   307,   922,    14,   199,
worker-08:             12,   400,   646,  1171,   307,   922,    14,   199,    12,   400]],
worker-08:        device='cuda:0')
localhost: epoch 0:
worker-02: ==============output================ tensor([[19706,    12,   745,   646,   411,  1171,   307,   922,    14,   199,
worker-02:             12,   400,   646,  1171,   307,   922,    14,   199,    12,   400]],
worker-02:        device='cuda:0')
worker-15: ==============output================ tensor([[19706,    12,   745,   646,   411,  1171,   307,   922,    14,   199,
worker-15:             12,   400,   646,  1171,   307,   922,    14,   199,    12,   400]],
worker-15:        device='cuda:0')
worker-09: ==============output================ tensor([[19706,    12,   745,   646,   411,  1171,   307,   922,    14,   199,
worker-09:             12,   400,   646,  1171,   307,   922,    14,   199,    12,   400]],
worker-08: epoch 0:
worker-09:        device='cuda:0')
worker-03: ==============output================ tensor([[19706,    12,   745,   646,   411,  1171,   307,   922,    14,   199,
worker-03:             12,   400,   646,  1171,   307,   922,    14,   199,    12,   400]],
worker-03:        device='cuda:0')
worker-02: epoch 0:
worker-15: epoch 0:
worker-01: ==============output================ tensor([[19706,    12,   745,   646,   411,  1171,   307,   922,    14,   199,
worker-01:             12,   400,   646,  1171,   307,   922,    14,   199,    12,   400]],
worker-01:        device='cuda:0')
`

### Expected behavior

I expected this code to load parts of the model on all GPUs, thus having less than 164MB on each GPUs and then do the inference of a single prompt using all part on all machines and giving out one response. This will be generalized to big models that can't fit on a single GPU.
